tests/runtime/abi.ts

```diff
@@ -28,4 +61,41 @@ export class Abi {
     const bytes = this.heap.subarray(start, start + len * 2)
     return this.text.decoder.decode(bytes)
   }
+
+  public writeArrayBuffer(value: ArrayBuffer): Pointer {
+    const totalSize = 8 + value.byteLength
+
+    // NOTE: AssemblyScript arena allocator requires allocs to be powers of 2
+    const allocSize = nextPowerOfTwo(totalSize)
+    const ptr = this.allocate(allocSize)
+
+    this.view.setUint32(ptr, value.byteLength, LE)
+    this.heap.set(new Uint8Array(value), ptr + 8)
+
+    return ptr
+  }
+
+  public writeBigInt(value: bigint): Pointer {
+    const bytes = toBytesLE(value)
+    const bufferPtr = this.writeArrayBuffer(bytes)
+
+    const ptr = this.allocate(16)
+    this.view.setUint32(ptr, bufferPtr, LE)
+    this.view.setUint32(ptr + 4, 0, LE)
+    this.view.setUint32(ptr + 8, bytes.length, LE)
+
+    return ptr
+  }
+}
+
+function nextPowerOfTwo(value: number): number {
+  value = ~~value // Convert to u32
+  value -= 1
+  value |= value >> 1
+  value |= value >> 2
+  value |= value >> 4
+  value |= value >> 8
+  value |= value >> 16
+  value += 1
+  return value
```

Comment:
I find this logic quite complex to understand how and why it works. Is this a standard approach that we could link in the comments? Otherwise why not just round op the `log2(value)`?

Reply:
This is a trick for counting leading zeros :stuck_out_tongue:. I left it in as a fun tidbit.